lib/rspec/expectations/fail_with.rb

```diff
@@ -19,18 +19,23 @@ def differ
 
       # Raises an RSpec::Expectations::ExpectationNotMetError with message.
       # @param [String] message
-      # @param [Object] expected
-      # @param [Object] actual
+      # @param [Object] matcher
       #
-      # Adds a diff to the failure message when `expected` and `actual` are
+      # Adds a diff to the failure message when `matcher.expected` and `matcher.actual` are
       # both present.
-      def fail_with(message, expected=nil, actual=nil)
+      def fail_with(message, matcher=nil)
         unless message
           raise ArgumentError, "Failure message is nil. Does your matcher define the " \
                                "appropriate failure_message[_when_negated] method to return a string?"
         end
 
-        message = ::RSpec::Matchers::ExpectedsForMultipleDiffs.from(expected).message_with_diff(message, differ, actual)
+        message = if !matcher.nil?
+                    differ_in_use = matcher.respond_to?(:differ) ? matcher.differ : differ
+
+                    ::RSpec::Matchers::ExpectedsForMultipleDiffs.from(matcher.expected).message_with_diff(message, differ_in_use, matcher.actual)
+                  else
+                    ::RSpec::Matchers::ExpectedsForMultipleDiffs.from(nil).message_with_diff(message, differ, nil)
```

Comment:
I'm yeah nahing this, expected and actual need to be passed in, differ could be an optional extra, this would eliminate the need for this if.